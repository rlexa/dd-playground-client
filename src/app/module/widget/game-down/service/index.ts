export * from './game-down.service';
