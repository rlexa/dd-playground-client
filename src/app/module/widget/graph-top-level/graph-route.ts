export enum GraphRoute {
  Walker = 'walker',
}
