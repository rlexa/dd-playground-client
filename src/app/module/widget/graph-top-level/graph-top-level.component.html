<app-navigation-content></app-navigation-content>
