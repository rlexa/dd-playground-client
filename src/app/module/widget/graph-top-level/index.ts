export * from './graph-top-level.module';
