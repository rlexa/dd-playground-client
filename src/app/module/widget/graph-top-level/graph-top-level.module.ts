import {CommonModule} from '@angular/common';
import {HttpClientModule} from '@angular/common/http';
import {NgModule} from '@angular/core';
import {RouterModule, Routes} from '@angular/router';
import {ROUTE_ROOT, ROUTE_WALKER, ROUTE_WILDCARD} from 'src/app/routing';
import {RoutedContentModule} from '../routed-content';
import {GraphTopLevelComponent} from './graph-top-level.component';

const ROUTING: Routes = [
  {
    path: ROUTE_ROOT,
    component: GraphTopLevelComponent,
    children: [
      {
        path: ROUTE_WALKER,
        loadChildren: () => import('src/app/module/widget/graph-walker/graph-walker.module').then((m) => m.GraphWalkerModule),
      },
      {path: ROUTE_ROOT, redirectTo: ROUTE_WALKER, pathMatch: 'full'},
      {path: ROUTE_WILDCARD, redirectTo: ROUTE_WALKER},
    ],
  },
];

@NgModule({
  imports: [CommonModule, HttpClientModule, RoutedContentModule, RouterModule.forChild(ROUTING)],
  exports: [GraphTopLevelComponent],
  declarations: [GraphTopLevelComponent],
})
class GraphTopLevelModule {}

export {GraphTopLevelModule, GraphTopLevelComponent};
