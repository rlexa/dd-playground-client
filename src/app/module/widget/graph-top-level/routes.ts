import {Route} from '@angular/router';
import {RouteRoot, RouteWild} from 'src/app/routing';
import {NavigationBarItem, NavigationBarItemsData} from '../navigation-bar';
import {GraphRoute} from './graph-route';
import {GraphTopLevelComponent} from './graph-top-level.component';

const routeNavs: Record<GraphRoute, NavigationBarItem> = {
  [GraphRoute.Walker]: {icon: 'items1', route: GraphRoute.Walker, label: 'Walker'},
};

const data: NavigationBarItemsData = {
  navigationBarItems: Object.values(GraphRoute).map((ii) => routeNavs[ii]),
};

export default [
  {
    path: RouteRoot,
    component: GraphTopLevelComponent,
    data,
    children: [
      {path: GraphRoute.Walker, loadChildren: () => import('../../feature/graph-walker/routes')},
      {path: RouteRoot, redirectTo: GraphRoute.Walker, pathMatch: 'full'},
      {path: RouteWild, redirectTo: GraphRoute.Walker},
    ],
  },
] as Route[];
