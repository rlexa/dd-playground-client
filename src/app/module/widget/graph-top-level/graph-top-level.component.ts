import { HttpClient } from '@angular/common/http';
import { ChangeDetectionStrategy, Component, OnInit } from '@angular/core';
import { GraphskyService, IGraphskyData, IGraphskyLinkRequest } from 'app/module/service/graphsky-api';
import { TAG_TYPE } from 'app/module/widget/graph-walker';
import { map, tap } from 'rxjs/operators';

interface TypedNode extends IGraphskyData {
  [TAG_TYPE]: string;
}

const csvToObjects = (type: string, csv: string, separator = ';') => {
  const ret = <TypedNode[]>[];
  const lines = csv.split('\n').map(ii => ii.trim()).filter(ii => ii.length);
  if (lines.length > 1) {
    const cols = lines.splice(0, 1)[0].split(separator).map(ii => ii.trim());
    lines.forEach(line => ret.push(
      line.split(separator).reduce((acc, ii, index) => ({ ...acc, [cols[index]]: ii }), { _type: type })));
  }
  return ret;
}

const csvToLinks = (csv: string, separator = ';', fromPrefix = 'from-', toPrefix = 'to-') => {
  const ret = <IGraphskyLinkRequest[]>[];
  const lines = csv.split('\n').map(ii => ii.trim()).filter(ii => ii.length);
  if (lines.length > 1) {
    const cols = lines.splice(0, 1)[0].split(separator).map(ii => ii.trim());
    lines.forEach(line => ret.push(
      line.split(separator).reduce((acc, ii, index) => {
        const col = cols[index];
        if (col.startsWith(fromPrefix)) {
          acc.from[col.substr(fromPrefix.length)] = ii;
        } else if (col.startsWith(toPrefix)) {
          acc.to[col.substr(toPrefix.length)] = ii;
        } else {
          acc.data[col] = ii;
        }
        return acc;
      }, <IGraphskyLinkRequest>{ fromNotExact: true, toNotExact: true, data: {}, from: {}, to: {} })));
  }
  return ret;
}

@Component({
  selector: 'app-graph-top-level',
  templateUrl: './graph-top-level.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush,
  providers: [GraphskyService],
})
export class GraphTopLevelComponent implements OnInit {
  constructor(
    private readonly graphsky: GraphskyService,
    private readonly http: HttpClient,
  ) { }

  ngOnInit() {
    [
      'country',
      'institution',
    ].forEach(source => this.http
      .get(`/assets/graph-data/${source}.csv`, { responseType: 'text', headers: { 'content-type': 'text/plain;charset=utf-8' } })
      .pipe(map(csv => csvToObjects(source, csv)), tap(items => console.log(`...adding ${items.length} items of: ${source}`)))
      .subscribe(items => this.graphsky.add(items)));

    [
      'country-institution',
    ].forEach(source => this.http
      .get(`/assets/graph-data/link-${source}.csv`, { responseType: 'text', headers: { 'content-type': 'text/plain;charset=utf-8' } })
      .pipe(map(csv => csvToLinks(csv)), tap(items => console.log(`...adding ${items.length} links of: ${source}`)))
      .subscribe(items => this.graphsky.link(items)));
  }
}
