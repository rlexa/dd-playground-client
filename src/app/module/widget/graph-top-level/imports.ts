import {CommonModule} from '@angular/common';
import {HttpClientModule} from '@angular/common/http';
import {RouterModule} from '@angular/router';
import {RoutedContentModule} from 'src/app/module/widget/routed-content';

export const imports = [CommonModule, HttpClientModule, RouterModule, RoutedContentModule];
