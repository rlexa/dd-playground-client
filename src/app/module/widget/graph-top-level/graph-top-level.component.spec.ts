import {HttpClient} from '@angular/common/http';
import {async, TestBed} from '@angular/core/testing';
import {RouterTestingModule} from '@angular/router/testing';
import {MockComponents} from 'ng-mocks';
import {of} from 'rxjs';
import {detectChanges, getByType, mockAll, overrideForChangeDetection} from 'src/app/test';
import {Mock} from 'ts-mockery';
import {GraphskyService} from '../../service/graphsky-api';
import {RoutedContentComponent} from '../routed-content';
import {GraphTopLevelComponent} from './graph-top-level.component';

describe('GraphTopLevelComponent', () => {
  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [RouterTestingModule],
      declarations: [GraphTopLevelComponent, MockComponents(RoutedContentComponent)],
      providers: [mockAll(GraphskyService), mockAll(HttpClient)],
    })
      .overrideComponent(GraphTopLevelComponent, overrideForChangeDetection)
      .compileComponents();
  }));

  beforeEach(() => {
    Mock.extend(getByType(HttpClient)).with({get: () => of(null)});
  });

  test('is created', () => {
    const fixture = TestBed.createComponent(GraphTopLevelComponent);
    detectChanges(fixture);
    expect(fixture).toMatchSnapshot();
  });
});
