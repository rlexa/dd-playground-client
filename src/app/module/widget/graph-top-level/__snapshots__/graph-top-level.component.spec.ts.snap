// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphTopLevelComponent is created 1`] = `
<app-graph-top-level
  __ngContext__={[Function LRootView]}
  graphsky={[Function GraphskyService]}
  http={[Function Object]}
>
  <app-navigation-content />
</app-graph-top-level>
`;
